server/src/main/java/org/elasticsearch/common/util/ReleasableIntArray.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+class ReleasableIntArray implements IntArray {
+    private final ReleasableBytesReference ref;
+
+    ReleasableIntArray(StreamInput in) throws IOException {
+        ref = in.readReleasableBytesReference();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBytesReference(ref);
+    }
+
+    @Override
+    public long size() {
+        return ref.length() / 4;
+    }
+
+    @Override
+    public int get(long index) {
+        if (index > Integer.MAX_VALUE / 4) {
+            // We can't serialize messages longer than 2gb anyway
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        return ref.getIntLE((int) index * 4);
+    }
+
+    @Override
+    public int set(long index, int value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int increment(long index, int inc) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void fill(long fromIndex, long toIndex, int value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void set(long index, byte[] buf, int offset, int len) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        /*
+         * We really have no idea how much heap we're using. We have a pretty
+         * good idea how much of the underlying buffer we're using, but that
+         * is already accounted otherwise.
+         */
+        return 0;
```

Comment:
NIT: I guess you could create a constant from `RamUsageEstimator.shallowSizeOfInstance(ReleasableIntArray.class)` and return that here. It doesn't really matter much, but it's the same number of lines as the above comment and slightly more accurate :)